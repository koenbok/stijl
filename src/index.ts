import * as React from "react";
import { css as glamor } from "glamor";

import { renderToString } from "react-dom/server";
import { renderStatic } from "glamor/server";

// TODO: Warn if there is a parent styling component through contexts
// TODO: Add cool types for media queries
// TODO: Add more psuedo css helpers

declare module "react" {
  interface HTMLAttributes<T> {
    css?: StylingRule;
  }

  interface SVGAttributes<T> {
    css?: StylingRule;
  }
}

const patchCreateElement = () => {
  const _createElement = React.createElement;

  const createElement = (...args) => {
    const props = Array.isArray(args) ? args[1] : args["props"];

    if (props && props.css && props.css instanceof StylingRule) {
      Object.assign(props, props.css.css());
      delete props["css"];
    }

    return _createElement.apply(this, args);
  };

  (React as any).createElement = createElement;
};

patchCreateElement();

export class StylingRule {
  private _css = {};

  constructor(css: React.CSSProperties) {
    this._css = css;
  }

  css() {
    return glamor(this._css) as any;
  }

  mediaQuery(query: string, css: React.CSSProperties) {
    this._css[`@media only screen and (${query})`] = css;
    return this;
  }

  hover(css: React.CSSProperties) {
    this._css[":hover"] = css;
    return this;
  }

  active(css: React.CSSProperties) {
    this._css[":active"] = css;
    return this;
  }

  placeholder(css: React.CSSProperties) {
    this._css[":placeholder"] = css;
    return this;
  }

  before(css: React.CSSProperties) {
    this._css[":before"] = css;
    return this;
  }
}

export const Style = (css: React.CSSProperties) => new StylingRule(css);

export const Styling = (props: { body: any }) => {
  let { html, css, ids } = renderStatic(() => renderToString(props.body));
  return React.createElement("style", {
    dangerouslySetInnerHTML: { __html: css }
  });
};
