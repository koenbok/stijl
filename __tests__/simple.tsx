import "jest";
import * as React from "react";
import * as ReactDOMServer from "react-dom/server";
import * as renderer from "react-test-renderer";
import { Style, Styling } from "../src/index";

const Template = props => {
  return (
    <html>
      <head>
        <Styling body={props.children} />
      </head>
      <body>{props.children}</body>
    </html>
  );
};

test("renders using spread", () => {
  const style = Style({ border: "1px solid red" });

  const tree = renderer
    .create(
      <Template>
        <div {...style.css()} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("renders using css property", () => {
  const style = Style({ border: "1px solid red" });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("renders hovers", () => {
  const style = Style({ border: "1px solid red" });

  style.hover({
    border: "1px solid green"
  });

  const tree = renderer
    .create(
      <Template>
        <div css={style} />
      </Template>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("update", () => {
  const style = Style({ border: "1px solid red" });

  style.update({ color: "red" });

  expect(style.style).toEqual({
    border: "1px solid red",
    color: "red"
  });

  expect(style).toBe(style);
});

test("merge", () => {
  const styleA = Style({ border: "1px solid red" });
  const styleB = styleA.merge({ color: "red" });

  expect(styleA.style).toEqual({
    border: "1px solid red"
  });

  expect(styleB.style).toEqual({
    border: "1px solid red",
    color: "red"
  });
});

test("copy", () => {
  const styleA = Style({ border: "1px solid red" });
  const styleB = styleA.copy();

  expect(styleA.style).toEqual({
    border: "1px solid red"
  });

  expect(styleB.style).toEqual({
    border: "1px solid red"
  });

  expect(styleA).not.toBe(styleB);
});

test("::placeholder", () => {
  const styleA = Style({});
  styleA.placeholder({ color: "red" });

  expect(styleA.style).toEqual({
    "::placeholder": { color: "red" }
  });
});
