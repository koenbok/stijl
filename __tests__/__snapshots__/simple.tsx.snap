// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders hovers 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1iz9137,[data-css-1iz9137] {border: 1px solid red;}.css-1iz9137:hover,[data-css-1iz9137]:hover {border: 1px solid green;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-1iz9137=""
    />
  </body>
</html>
`;

exports[`renders using css property 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1pe0egf,[data-css-1pe0egf] {border: 1px solid red;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-1pe0egf=""
    />
  </body>
</html>
`;

exports[`renders using spread 1`] = `
<html>
  <head>
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1pe0egf,[data-css-1pe0egf] {border: 1px solid red;}",
        }
      }
    />
  </head>
  <body>
    <div
      data-css-1pe0egf=""
    />
  </body>
</html>
`;
